ext/vulnsrc/redhat/redhat.go

```diff
@@ -0,0 +1,684 @@
+}
+
+// get advisories from the given oval document which were issued since the last update (based on db value)
+func ProcessAdvisoriesSinceLastDbUpdate(ovalDoc OvalV2Document, datastore database.Datastore) ([]ParsedAdvisory, error) {
+	sinceDate := DbLookupLastAdvisoryDate(datastore)
+	var advisories []ParsedAdvisory
+	for _, definition := range ovalDoc.DefinitionSet.Definitions {
+		// check whether this is a supported definition type
+		if (!IsSupportedDefinitionType(definition.Class)) {
+			// not supported; skip it
+			continue
+		}
+		// check if this entry has already been processed (based on its issued date)
+		if IsAdvisorySinceDate(sinceDate, definition.Metadata.Advisory.Issued.Date) {
+			// this advisory was issued since the last advisory date in the database; add it
+			// debug
+			log.Info(fmt.Sprintf("Found advisory issued since the last known advisory date (%s) in database: %s (%s)",
+				sinceDate,
+				definition.Metadata.Title, definition.Metadata.Advisory.Issued.Date))
+			advisories = append(advisories, ParseAdvisory(definition, ovalDoc))
+		} else if IsAdvisorySameDate(sinceDate, definition.Metadata.Advisory.Issued.Date) {
+			parsedAdvisory := ParseAdvisory(definition, ovalDoc)
+			// advisory date is coarse (YYYY-MM-dd format) date only,
+			// so it's possible that we'll see an advisory multiple times within the same day;
+			// check the db in this case to be sure
+			if (!DbLookupIsAdvisoryProcessed(parsedAdvisory, datastore)) {
+				// this advisory id/version hasn't been processed yet; add it
+				// debug
+				log.Info(fmt.Sprintf("Found unprocessed advisory issued on the last known advisory date (%s) in database: %s (%s)",
+					sinceDate,
+					definition.Metadata.Title, definition.Metadata.Advisory.Issued.Date))
+				advisories = append(advisories, parsedAdvisory)
+			}
+		} else {
+			// this advisory was issued before the last advisory date in the database, so already processed; skip it
+			// debug
+			log.Info(fmt.Sprintf("Skipping advisory issued before the last known advisory date (%s) in database: %s (%s)",
+				sinceDate,
+				definition.Metadata.Title, definition.Metadata.Advisory.Issued.Date))
+		}
+	}
+	// debug-only info
+	out, _ := xml.MarshalIndent(ovalDoc, " ", "  ")
+	log.Debug(string(out))
+
+	return advisories, nil
+}
+
+func ParseAdvisory(definition OvalV2AdvisoryDefinition, ovalDoc OvalV2Document) (ParsedAdvisory) {
+	parsedAdvisory := ParsedAdvisory{
+		Class: definition.Class,
+		Id: definition.Id,
+		Version: definition.Version,
+		Metadata: definition.Metadata,
+		Criteria: definition.Criteria,
+		PackageList: GetPackageList(definition.Criteria, ovalDoc),
+	}
+	return parsedAdvisory
+}
+
+func GetPackageList(criteria OvalV2Criteria, ovalDoc OvalV2Document) (parsedNvras []ParsedRmpNvra) {
+	criterions := extractAllCriterions(criteria)
+	for _, criterion := range criterions {
+		// get package info
+		parsedNvras = append(parsedNvras, FindPackageNvraInfo(criterion.TestRef, ovalDoc))
```

Comment:
The reason why there are so many warning logs that I showed you yesterday is that you create a feature object based on criterion that is not relevant. I tried to debug it today and found out that you create features based on:
`<criterion comment="atomic-openshift-sdn-ovs is signed with Red Hat redhatrelease2 key" test_ref="oval:com.redhat.rhsa:tst:20152544002"/>`

If you follow a `test_ref` for a given criterion you don't get any EVR nor architecture. You should filter out those criteria.